File: new/python/121_best_time_to_buy_and_sell_stock_two_pointers.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # idea: two pointer
        # track buying and selling day using left and right pointers
        # left = buying day
        # right = selling day
        # if left < right, update max profits
        # if not make left point to right, because buying day is not less than selling day, so our selling day is lte, so we make that
        # our new buying day
        # move the selling day to the next day regardless
        # do this until we reach the end
        l_buying_day = 0
        r_selling_day = 1
        current_profits = 0

        while r_selling_day < len(prices):
            buy = prices[l_buying_day]
            sell = prices[r_selling_day]
            if buy < sell:
                current_profits = max(current_profits, sell - buy)
            else:
                l_buying_day = r_selling_day
            r_selling_day += 1

        return current_profits
        # time: O(n)
        # space: O(1)

File: new/cpp/121_best_time_to_buy_and_sell_stock_two_pointers.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // idea: two pointers
        // left = buy day
        // right = sell day
        // find profits for all sell (right) > buy (left) as long as right < prices.size()
        // if right is not gt left, then make left = right
        // move right to next each iteration
        int left_buy_day = 0;
        int right_sell_day = 1;
        int profits = 0;

        while (right_sell_day < prices.size()) {
            int buy = prices[left_buy_day];
            int sell = prices[right_sell_day];
            if (buy < sell) {
                profits = max(profits, sell - buy);
            } else {
                left_buy_day = right_sell_day;
            }
            right_sell_day += 1;
        }

        return profits;
    }

    // time: O(n)
    // space: O(1)
};
